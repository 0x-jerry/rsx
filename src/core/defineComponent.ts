import { isFn } from '@0x-jerry/utils'
import type { Merge } from 'type-fest'
import type { ToMaybeRef } from '.'
import type { DefineProps } from './props'

export interface PropOption<T = any> {
  type?: T
  default?: T | (() => T)
}

export type PropsType<T = any> = Record<string, T | PropOption<T>>

export interface ComponentOption<Props extends PropsType = {}> {
  props?: Props
  toRefs?: boolean
}

interface CommonProps {
  class?: any
  [key: string]: any
}

export type FunctionalComponent<P extends PropsType = {}> = (
  props: Merge<P, CommonProps>,
  children?: any[],
) => JSX.Element

/**
 * without option
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * with option
 * @param opt
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P>,
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * implement
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P> | FunctionalComponent,
  impl?: FunctionalComponent,
): FunctionalComponent {
  if (isFn(opt)) return opt

  // todo check props type
  return impl!
}

export const dc = defineComponent
