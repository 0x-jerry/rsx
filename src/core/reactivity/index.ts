export { computed, reactive, ref } from '@vue/reactivity'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './utils'
export * from './watch'
