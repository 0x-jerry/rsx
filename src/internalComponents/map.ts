import { type Ref, shallowRef } from '@vue/reactivity'
import { asyncWatcherScheduler } from '@/reactivity/scheduler'
import { defineComponentName } from '@/test'
import {
  createAnchorNode,
  dispatchAnchorMovedEvent,
  getAnchorFirstChildNode,
  isAnchorNode,
  listenAnchorMoveEvent,
  setAnchorNodeFirstChildren,
} from '../anchorNode'
import { type DNodeContext, runWithContext } from '../context'
import {
  defineComponent,
  type ExposedFunctionalComponent,
  type FunctionalComponent,
} from '../defineComponent'
import { mount, onBeforeMount, unmount, useContext, useWatch } from '../hook'
import { insertBefore } from '../nodeOp'
import { type ComponentNode, createComponentNode } from '../nodes/ComponentNode'
import { computed } from '../reactivity'

export interface MapItemProps<Item> {
  item: Item
  index: number
}

export type MapItemComponent<Item> = ExposedFunctionalComponent<
  MapItemProps<Item>
>

export interface MapComponentProps<T> {
  list: T[]
  key?: (item: T, index: number) => unknown
  render: FunctionalComponent<MapItemProps<T>>
}

interface ChildContext extends ComponentNode {
  /**
   * mark this is a reuse element
   */
  _r?: boolean
  _props?: {
    item: Ref<unknown>
    index: Ref<number>
  }
}

export const VMap = defineComponent(<T>(props: MapComponentProps<T>) => {
  const ctx = useContext()

  const anchorNode = createAnchorNode('map')

  let children: ChildContext[] = []

  let dataContextMap = new Map<unknown, ChildContext[]>()

  const childrenKeys = computed(() =>
    props.list.map((item, idx) => getItemKey(item, idx)),
  )

  useWatch(
    childrenKeys,
    () => {
      runWithContext(update, ctx)
    },
    {
      scheduler: asyncWatcherScheduler,
    },
  )

  onBeforeMount(() => {
    runWithContext(update, ctx)
  })

  listenAnchorMoveEvent(anchorNode, () => {
    children.forEach((child) => {
      const childEl = child.context?.el
      if (childEl) {
        insertBefore(anchorNode, childEl)

        dispatchAnchorMovedEvent(childEl)
      }
    })
  })

  return anchorNode

  function update() {
    const c1: ChildContext[] = children
    const c2: ChildContext[] = buildNewChildren()

    let i = 0
    const l2 = c2.length
    let e1 = c1.length - 1
    let e2 = l2 - 1

    while (i <= e1 && i <= e2) {
      const n1 = c1[i]
      const n2 = c2[i]

      if (n1 !== n2) {
        break
      }
      n1._r = false

      i++
    }

    while (i <= e1 && i <= e2) {
      const n1 = c1[e1]
      const n2 = c2[e2]
      if (n1 !== n2) {
        break
      }
      n1._r = false

      e1--
      e2--
    }

    if (i > e1) {
      while (i <= e2) {
        const n = c2[i]
        const anchor = getFirstChildOfNode(c1[e1 + 1]?.context) || anchorNode

        const nEl = n.context.el
        if (nEl) {
          insertBefore(anchor, nEl)
          dispatchAnchorMovedEvent(nEl)
        }

        if (ctx._mounted) {
          mount(n.context)
        }

        i++
      }
    } else if (i > e2) {
      // already unmounted
      // no need to do anything
    } else {
      const s1 = i
      const s2 = i

      const newSequence: number[] = []
      const oldToNew = new Map<number, number>()

      const cc1 = c1.slice(s1, e1 + 1)

      // todo, move check
      for (let j = s2; j <= e2; j++) {
        const element = c2[j]
        const oldIdx = cc1.indexOf(element)
        if (oldIdx !== -1) {
          newSequence.push(oldIdx)
          oldToNew.set(oldIdx, j)
        }
      }

      const increasingNewIndexSequence = getSequence(newSequence)

      let anchorPreviousNode =
        c1[i - 1]?.context.el ||
        (c1[0]
          ? getFirstChildOfNode(c1[0]?.context)?.previousSibling
          : anchorNode.previousSibling)

      for (i = s2; i <= e2; i++) {
        const n2 = c2[i]

        if (
          increasingNewIndexSequence.length &&
          oldToNew.get(newSequence[increasingNewIndexSequence[0]]) === i
        ) {
          n2._r = false
          anchorPreviousNode = n2.context.el
          increasingNewIndexSequence.shift()
          continue
        }

        const n2El = n2.context.el
        if (n2El) {
          const anchor =
            (anchorPreviousNode
              ? anchorPreviousNode.nextSibling
              : c1[0]?.context.el?.parentElement
                ? getFirstChildOfNode(c1[0]?.context)
                : null) || anchorNode

          insertBefore(anchor, n2El)
          dispatchAnchorMovedEvent(n2El)

          anchorPreviousNode = n2El
        }

        if (n2._r) {
          n2._r = false
        } else {
          if (ctx._mounted) {
            mount(n2.context)
          }
        }
      }
    }

    children = c2

    // update first child
    {
      const firstEl = children.find((child) => child.context.el)?.context.el
      setAnchorNodeFirstChildren(anchorNode, firstEl)
    }
  }

  function buildNewChildren() {
    const newChildren: ChildContext[] = []

    const newDataContextMap = new Map<unknown, ChildContext[]>()

    props.list.forEach((item, idx) => {
      const dataKey = childrenKeys.value[idx]

      if (dataContextMap.has(dataKey)) {
        const reuseCtx = popItemFromMap(dataContextMap, dataKey)
        reuseCtx._r = true
        if (reuseCtx._props) {
          reuseCtx._props.item.value = item
          reuseCtx._props.index.value = idx
        }

        appendItemToMap(newDataContextMap, dataKey, reuseCtx)

        newChildren.push(reuseCtx)

        return
      }

      const childProps = {
        item: shallowRef(item),
        index: shallowRef(idx),
      }

      const newCtx = createComponentNode(
        props.render,
        childProps,
        [],
      ) as ChildContext

      newCtx._props = childProps

      newCtx.initialize()
      newCtx.context.name = 'VMap.item'

      appendItemToMap(newDataContextMap, dataKey, newCtx)
      newChildren.push(newCtx)
    })

    dataContextMap.values().forEach((ctxList) => {
      ctxList.forEach((child) => {
        if (child.context) {
          unmount(child.context)
        } else {
          throw new Error('child without instance')
        }
      })
    })

    dataContextMap = newDataContextMap

    return newChildren
  }

  function getItemKey(item: T, idx: number) {
    return props.key ? props.key(item, idx) : item
  }
})

defineComponentName(VMap, 'VMap')

function appendItemToMap<K, V>(map: Map<K, V[]>, key: K, value: V) {
  let list = map.get(key)
  if (!list) {
    list = []
    map.set(key, list)
  }

  list.push(value)
}

function popItemFromMap<K, V>(map: Map<K, V[]>, key: K) {
  // biome-ignore lint/style/noNonNullAssertion: already checked before use it
  const collection = map.get(key)!
  // biome-ignore lint/style/noNonNullAssertion: already checked before use it
  const item = collection.shift()!

  if (!collection.length) {
    map.delete(key)
  }

  return item
}

// https://en.wikipedia.org/wiki/Longest_increasing_subsequence
function getSequence(arr: number[]): number[] {
  const p = arr.slice()
  const result = [0]
  // biome-ignore lint/suspicious/noImplicitAnyLet: just ignore this
  let i, j, u, v, c
  const len = arr.length
  for (i = 0; i < len; i++) {
    const arrI = arr[i]
    if (arrI !== 0) {
      j = result[result.length - 1]
      if (arr[j] < arrI) {
        p[i] = j
        result.push(i)
        continue
      }
      u = 0
      v = result.length - 1
      while (u < v) {
        c = (u + v) >> 1
        if (arr[result[c]] < arrI) {
          u = c + 1
        } else {
          v = c
        }
      }
      if (arrI < arr[result[u]]) {
        if (u > 0) {
          p[i] = result[u - 1]
        }
        result[u] = i
      }
    }
  }
  u = result.length
  v = result[u - 1]
  while (u-- > 0) {
    result[u] = v
    v = p[v]
  }
  return result
}

function getFirstChildOfNode(node?: DNodeContext) {
  if (!node?.el) return

  if (isAnchorNode(node.el)) {
    return getAnchorFirstChildNode(node.el)
  }

  return node.el
}
