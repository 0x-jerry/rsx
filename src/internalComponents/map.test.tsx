import { dc } from '../defineComponent'
import { onBeforeMount, onMounted, onUnmounted } from '../hook'
import { $, nextTick, ref, toBindingRefs } from '../reactivity'
import { contextToJson, defineComponentName, mountTestApp } from '../test'
import { VMap } from './map'

describe('map component', () => {
  it('non-reactivity data', async () => {
    const App = dc(() => {
      const list = [1, 2, 3]
      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1', '2', '3'])
  })

  it('complex data array', async () => {
    const App = dc(() => {
      const list = [{ a: 1 }, { a: 2 }, { a: 3 }]

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '[object Object]',
      '[object Object]',
      '[object Object]',
    ])
  })

  it('reactivity data', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [4, 5, 6]
      })

      return (
        <div>
          <VMap
            list={list}
            render={(props) => <div class="item">{props.item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['4', '5', '6'])
  })

  it('reactivity data with index', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [2, 1, 3]
      })

      return (
        <div>
          <VMap
            list={list}
            render={(props) => {
              const { item, index } = toBindingRefs(props)

              return (
                <div class="item">
                  {item}-{index}
                </div>
              )
            }}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['2-0', '1-1', '3-2'])
  })

  it('custom keyed reactivity data', async () => {
    interface Item {
      name: string
      key: number
    }

    const App = dc(() => {
      const list = ref<Item[]>([
        {
          name: '1',
          key: 1,
        },
        {
          name: '2',
          key: 2,
        },
        {
          name: '3',
          key: 3,
        },
      ])

      onMounted(() => {
        list.value = [
          {
            name: '2',
            key: 1,
          },
          {
            name: '1',
            key: 2,
          },
          {
            name: '4',
            key: 3,
          },
        ]
      })

      return (
        <div>
          <VMap
            list={list}
            key={(item) => item.key}
            render={(props) => {
              return <div class="item">{$(() => props.item.name)}</div>
            }}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['2', '1', '4'])
  })

  it('reuse node', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3, 4])

      onMounted(() => {
        list.value = [4, 2, 3, 1]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['4', '2', '3', '1'])
  })

  it('add item to list', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [1, 4, 2, 3]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1', '4', '2', '3'])
  })

  it('add item from start', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.unshift(4)
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['4', '1', '2', '3'])
  })

  it('add item to the end', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.push(4)
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1', '2', '3', '4'])
  })

  it('delete item to list', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [1, 3]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1', '3'])
  })

  it('delete item from start', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.shift()
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['2', '3'])
  })

  it('delete item from end', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.pop()
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => <div class="item">{item}</div>}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1', '2'])
  })
})

describe('map component with fragment', () => {
  it('non-reactivity data', async () => {
    const App = dc(() => {
      const list = [1, 2, 3]
      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}</div>
                <div class="item">{item}</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1', '1', '2', '2', '3', '3'])
  })

  it('complex data array', async () => {
    const App = dc(() => {
      const list = [{ a: 1 }, { a: 2 }, { a: 3 }]

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}</div>
                <div class="item">{item}</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '[object Object]',
      '[object Object]',
      '[object Object]',
      '[object Object]',
      '[object Object]',
      '[object Object]',
    ])
  })

  it('reactivity data', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [4, 5, 6]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}</div>
                <div class="item">{item}</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['4', '4', '5', '5', '6', '6'])
  })

  it('reuse node', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3, 4])

      onMounted(() => {
        list.value = [4, 2, 3, 1]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                <div class="item">{item}-1</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '4-0',
      '4-1',
      '2-0',
      '2-1',
      '3-0',
      '3-1',
      '1-0',
      '1-1',
    ])
  })

  it('reuse node 1', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3, 4])

      onMounted(() => {
        list.value = [4, 2, 3, 1]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                {/** biome-ignore lint/complexity/noUselessFragments: test purpose */}
                <>
                  <div class="item">{item}-1</div>
                </>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '4-0',
      '4-1',
      '2-0',
      '2-1',
      '3-0',
      '3-1',
      '1-0',
      '1-1',
    ])
  })

  it('add item to list', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [1, 4, 2, 3]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                <div class="item">{item}-1</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '1-0',
      '1-1',
      '4-0',
      '4-1',
      '2-0',
      '2-1',
      '3-0',
      '3-1',
    ])
  })

  it('add item from start', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.unshift(4)
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                <div class="item">{item}-1</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '4-0',
      '4-1',
      '1-0',
      '1-1',
      '2-0',
      '2-1',
      '3-0',
      '3-1',
    ])
  })

  it('add item to the end', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.push(4)
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                <div class="item">{item}-1</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '1-0',
      '1-1',
      '2-0',
      '2-1',
      '3-0',
      '3-1',
      '4-0',
      '4-1',
    ])
  })

  it('delete item to list', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [1, 3]
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                <div class="item">{item}-1</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1-0', '1-1', '3-0', '3-1'])
  })

  it('delete item from start', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.shift()
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                <div class="item">{item}-1</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['2-0', '2-1', '3-0', '3-1'])
  })

  it('delete item from end', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.pop()
      })

      return (
        <div>
          <VMap
            list={list}
            render={({ item }) => (
              <>
                <div class="item">{item}-0</div>
                <div class="item">{item}-1</div>
              </>
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1-0', '1-1', '2-0', '2-1'])
  })
})

describe('map with map', () => {
  it('non-reactivity data', async () => {
    const App = dc(() => {
      const list = [1, 2, 3]
      return (
        <div>
          <VMap
            list={list}
            render={(props) => (
              <VMap
                list={[1, 2]}
                render={(props1) => (
                  <>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-0`)}
                    </div>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-1`)}
                    </div>
                  </>
                )}
              />
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '1-1-0',
      '1-1-1',
      '1-2-0',
      '1-2-1',
      '2-1-0',
      '2-1-1',
      '2-2-0',
      '2-2-1',
      '3-1-0',
      '3-1-1',
      '3-2-0',
      '3-2-1',
    ])
  })

  it('reactivity data', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [4, 5, 6]
      })

      return (
        <div>
          <VMap
            list={list}
            render={(props) => (
              <VMap
                list={[1, 2]}
                render={(props1) => (
                  <>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-0`)}
                    </div>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-1`)}
                    </div>
                  </>
                )}
              />
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '4-1-0',
      '4-1-1',
      '4-2-0',
      '4-2-1',
      '5-1-0',
      '5-1-1',
      '5-2-0',
      '5-2-1',
      '6-1-0',
      '6-1-1',
      '6-2-0',
      '6-2-1',
    ])
  })

  it('add item to list', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [1, 4, 2, 3]
      })

      return (
        <div>
          <VMap
            list={list}
            render={(props) => (
              <VMap
                list={[1, 2]}
                render={(props1) => (
                  <>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-0`)}
                    </div>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-1`)}
                    </div>
                  </>
                )}
              />
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '1-1-0',
      '1-1-1',
      '1-2-0',
      '1-2-1',
      '4-1-0',
      '4-1-1',
      '4-2-0',
      '4-2-1',
      '2-1-0',
      '2-1-1',
      '2-2-0',
      '2-2-1',
      '3-1-0',
      '3-1-1',
      '3-2-0',
      '3-2-1',
    ])
  })

  it('add item from start', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.unshift(4)
      })

      return (
        <div>
          <VMap
            list={list}
            render={(props) => (
              <VMap
                list={[1, 2]}
                render={(props1) => (
                  <>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-0`)}
                    </div>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-1`)}
                    </div>
                  </>
                )}
              />
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '4-1-0',
      '4-1-1',
      '4-2-0',
      '4-2-1',
      '1-1-0',
      '1-1-1',
      '1-2-0',
      '1-2-1',
      '2-1-0',
      '2-1-1',
      '2-2-0',
      '2-2-1',
      '3-1-0',
      '3-1-1',
      '3-2-0',
      '3-2-1',
    ])
  })

  it('add item to the end', async () => {
    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value.push(4)
      })

      return (
        <div>
          <VMap
            list={list}
            render={(props) => (
              <VMap
                list={[1, 2]}
                render={(props1) => (
                  <>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-0`)}
                    </div>
                    <div class="item">
                      {$(() => `${props.item}-${props1.item}-1`)}
                    </div>
                  </>
                )}
              />
            )}
          />
        </div>
      )
    })

    const el = mountTestApp(App)

    await nextTick()

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql([
      '1-1-0',
      '1-1-1',
      '1-2-0',
      '1-2-1',
      '2-1-0',
      '2-1-1',
      '2-2-0',
      '2-2-1',
      '3-1-0',
      '3-1-1',
      '3-2-0',
      '3-2-1',
      '4-1-0',
      '4-1-1',
      '4-2-0',
      '4-2-1',
    ])
  })

  it('lifecycle hook', async () => {
    const lifecycle: string[] = []

    const Item = dc<{ item: number }>((props) => {
      onUnmounted(() => {
        lifecycle.push(`um-${props.item}`)
      })
      onBeforeMount(() => {
        lifecycle.push(`bm-${props.item}`)
      })
      onMounted(() => {
        lifecycle.push(`m-${props.item}`)
      })

      return <div>{props.item}</div>
    })

    const App = dc(() => {
      const list = ref([1, 2, 3])

      onMounted(() => {
        list.value = [3, 1, 4]
      })

      return <VMap list={list} render={Item} />
    })

    mountTestApp(App)
    await nextTick()

    expect(lifecycle).eql([
      'bm-1',
      'm-1',
      'bm-2',
      'm-2',
      'bm-3',
      'm-3',
      'um-2',
      'bm-4',
      'm-4',
    ])
  })
})

describe('map context tree', () => {
  it('static value', () => {
    const A = dc((_, children) => (
      <div class="A">
        <span>a</span>
        {children}
      </div>
    ))

    defineComponentName(A, 'A')

    const B = dc((_, children) => (
      <div class="B">
        <span>b</span>
        {children}
      </div>
    ))

    defineComponentName(B, 'B')

    const App = dc(() => (
      <A>
        <VMap list={[1, 2, 3]} render={(props) => <B>{props.item}</B>} />
      </A>
    ))

    defineComponentName(App, 'App')

    const root = mountTestApp(App)

    expect(root).toMatchSnapshot('html')

    const ctxTree = contextToJson(root._.context)

    expect(ctxTree).toMatchSnapshot('ctx tree')
  })

  it('reactive value', async () => {
    const A = dc((_, children) => (
      <div class="A">
        <span>a</span>
        {children}
      </div>
    ))

    defineComponentName(A, 'A')

    const B = dc((_, children) => (
      <div class="B">
        <span>b</span>
        {children}
      </div>
    ))

    defineComponentName(B, 'B')

    const App = dc(() => {
      const data = ref([1, 2, 3])

      onMounted(() => {
        data.value = [2, 1, 4, 3]
      })

      return (
        <A>
          <VMap list={data} render={(props) => <B>{props.item}</B>} />
        </A>
      )
    })

    defineComponentName(App, 'App')

    const root = mountTestApp(App)
    await nextTick()

    expect(root).toMatchSnapshot('html')

    const ctxTree = contextToJson(root._.context)

    expect(ctxTree).toMatchSnapshot('ctx tree')
  })
})
