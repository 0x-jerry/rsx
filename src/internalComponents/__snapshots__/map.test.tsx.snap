// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`map context tree > reactive value > ctx tree 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "name": "B",
                },
              ],
              "name": "VMap.item",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "B",
                },
              ],
              "name": "VMap.item",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "B",
                },
              ],
              "name": "VMap.item",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "B",
                },
              ],
              "name": "VMap.item",
            },
          ],
          "name": "VMap",
        },
      ],
      "name": "A",
    },
  ],
  "name": "Root",
}
`;

exports[`map context tree > reactive value > html 1`] = `
<div
  id="app"
>
  <div
    class="A"
  >
    <span>
      a
    </span>
    <div
      class="B"
    >
      <span>
        b
      </span>
      2
    </div>
    <div
      class="B"
    >
      <span>
        b
      </span>
      1
    </div>
    <div
      class="B"
    >
      <span>
        b
      </span>
      4
    </div>
    <div
      class="B"
    >
      <span>
        b
      </span>
      3
    </div>
    <!--map-->
  </div>
</div>
`;

exports[`map context tree > static value > ctx tree 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "name": "B",
                },
              ],
              "name": "VMap.item",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "B",
                },
              ],
              "name": "VMap.item",
            },
            {
              "children": [
                {
                  "children": [],
                  "name": "B",
                },
              ],
              "name": "VMap.item",
            },
          ],
          "name": "VMap",
        },
      ],
      "name": "A",
    },
  ],
  "name": "Root",
}
`;

exports[`map context tree > static value > html 1`] = `
<div
  id="app"
>
  <div
    class="A"
  >
    <span>
      a
    </span>
    <div
      class="B"
    >
      <span>
        b
      </span>
      1
    </div>
    <div
      class="B"
    >
      <span>
        b
      </span>
      2
    </div>
    <div
      class="B"
    >
      <span>
        b
      </span>
      3
    </div>
    <!--map-->
  </div>
</div>
`;
